### PR TITLE
8325169: Reduce String::indexOf overheads

### PR DESCRIPTION
**Pull request by cl4es**
This patch streamlines and specializes various `String::indexOf` methods. Mainly avoids the need for clamping and doing checks that are redundant in almost all cases, moving the checks to the API boundary where they are needed. 

This improves performance both at peak and during startup/warmup. Since indexOf is heavily used in bootstrapping code it makes sense to slightly dial back abstraction and delegation, which in this case also brought some benefit to peak performance.

Testing: tier1-3

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325169](https://bugs.openjdk.org/browse/JDK-8325169): Reduce String::indexOf overheads (**Enhancement** - P4)

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Raffaello Giulietti](https://openjdk.org/census#rgiulietti) (@rgiulietti - **Reviewer**) ⚠️ Review applies to [aad343b3](https://git.openjdk.org/jdk/pull/17685/files/aad343b3763ac2e788a6ed9a714e897e0db38e1c)
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17685/head:pull/17685` \
`$ git checkout pull/17685`

Update a local copy of the PR: \
`$ git checkout pull/17685` \
`$ git pull https://git.openjdk.org/jdk.git pull/17685/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17685`

View PR using the GUI difftool: \
`$ git pr show -t 17685`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17685.diff">https://git.openjdk.org/jdk/pull/17685.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17685#issuecomment-1924052321)

### CODE REVIEW

**Comment by cl4es** [comment]
M1 MacBook:
```Name                                              Cnt   Base   Error    Test   Error  Unit  Change
StringIndexOf.searchChar16LongSuccess               5 14,909 ± 3,284  14,480 ± 0,256 ns/op   1,03x (p = 0,324 )
StringIndexOf.searchChar16LongWithOffsetSuccess     5 14,715 ± 0,870  14,373 ± 0,382 ns/op   1,02x (p = 0,024 )
StringIndexOf.searchChar16MediumSuccess             5  6,611 ± 0,016   6,476 ± 0,050 ns/op   1,02x (p = 0,000*)
StringIndexOf.searchChar16MediumWithOffsetSuccess   5  7,151 ± 0,026   6,884 ± 0,014 ns/op   1,04x (p = 0,000*)
StringIndexOf.searchChar16ShortSuccess              5  2,667 ± 0,007   2,369 ± 0,005 ns/op   1,13x (p = 0,000*)
StringIndexOf.searchChar16ShortWithOffsetSuccess    5  2,494 ± 0,112   2,102 ± 0,006 ns/op   1,19x (p = 0,000*)
StringIndexOf.searchCharLongSuccess                 5  5,995 ± 0,017   5,276 ± 0,017 ns/op   1,14x (p = 0,000*)
StringIndexOf.searchCharLongWithOffsetSuccess       5  6,377 ± 0,118   5,735 ± 0,014 ns/op   1,11x (p = 0,000*)
StringIndexOf.searchCharMediumSuccess               5  2,350 ± 0,013   1,927 ± 0,012 ns/op   1,22x (p = 0,000*)
StringIndexOf.searchCharMediumWithOffsetSuccess     5  2,675 ± 0,002   2,247 ± 0,057 ns/op   1,19x (p = 0,000*)
StringIndexOf.searchCharShortSuccess                5  1,719 ± 0,002   1,245 ± 0,005 ns/op   1,38x (p = 0,000*)
StringIndexOf.searchCharShortWithOffsetSuccess      5  1,635 ± 0,006   1,247 ± 0,009 ns/op   1,31x (p = 0,000*)
  * = significant
```

`-Xint`, M1 MacBook:
```Name                                              Cnt     Base    Error      Test    Error  Unit  Change
StringIndexOf.searchChar16LongSuccess               5 4683,186 ±  6,431  4482,179 ± 26,216 ns/op   1,04x (p = 0,000*)
StringIndexOf.searchChar16LongWithOffsetSuccess     5 4586,540 ± 39,266  4403,976 ± 18,652 ns/op   1,04x (p = 0,000*)
StringIndexOf.searchChar16MediumSuccess             5 2123,355 ±  3,885  1942,100 ± 77,042 ns/op   1,09x (p = 0,000*)
StringIndexOf.searchChar16MediumWithOffsetSuccess   5 2007,646 ± 52,087  1859,545 ± 67,424 ns/op   1,08x (p = 0,000*)
StringIndexOf.searchChar16ShortSuccess              5  548,599 ±  3,048   386,671 ± 60,860 ns/op   1,42x (p = 0,000*)
StringIndexOf.searchChar16ShortWithOffsetSuccess    5  459,164 ±  5,782   302,377 ± 39,086 ns/op   1,52x (p = 0,000*)
StringIndexOf.searchCharLongSuccess                 5  908,953 ± 78,580   799,114 ± 34,290 ns/op   1,14x (p = 0,000*)
StringIndexOf.searchCharLongWithOffsetSuccess       5  855,726 ± 28,357   790,690 ± 19,452 ns/op   1,08x (p = 0,000*)
StringIndexOf.searchCharMediumSuccess               5  360,903 ±  3,214   262,320 ±  1,706 ns/op   1,38x (p = 0,000*)
StringIndexOf.searchCharMediumWithOffsetSuccess     5  322,950 ±  5,790   268,992 ±  8,008 ns/op   1,20x (p = 0,000*)
StringIndexOf.searchCharShortSuccess                5  274,718 ±  7,208   174,274 ±  3,445 ns/op   1,58x (p = 0,000*)
StringIndexOf.searchCharShortWithOffsetSuccess      5  254,141 ± 11,992   186,917 ± 11,471 ns/op   1,36x (p = 0,000*)
  * = significant
```

File: src/java.base/share/classes/java/lang/String.java [hunk 1]
```diff
@@ -2500,8 +2501,9 @@ public int indexOf(int ch) {
      * {@code fromIndex} were larger than the string length, or were negative.
      */
     public int indexOf(int ch, int fromIndex) {
-        return isLatin1() ? StringLatin1.indexOf(value, ch, fromIndex, length())
-                : StringUTF16.indexOf(value, ch, fromIndex, length());
+        fromIndex = Math.max(0, fromIndex);
+        return isLatin1() ? StringLatin1.indexOf(value, ch, fromIndex, value.length)
+                : StringUTF16.indexOf(value, ch, fromIndex, value.length >> 1);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/lang/String.java:2506]: This needs to include the check for `fromIndex >= this.length()`:
```suggestion
        fromIndex = Math.max(0, fromIndex);
        int toIndex = length();
        if (fromIndex >= toIndex) {
            return -1;
        }
        return isLatin1()
                ? StringLatin1.indexOf(value, ch, fromIndex, toIndex)
                : StringUTF16.indexOf(value, ch, fromIndex, toIndex);
```

**Review by RogerRiggs — APPROVED** [review]
Looks ok as a performance improvement; some comments about assertions might be useful for future readers.

**Comment by cl4es** [comment]
Thanks for looking at this!

I want to do a thorough check of all the intrinsics myself before I feel comfortable integrating. Tests should cover most out-of-bounds scenarios, but I'll double-check that as well.

**Review by rgiulietti — APPROVED** [review]
LGTM

File: src/java.base/share/classes/java/lang/String.java (hunk 1 above)
Comment (cl4es) [reply src/java.base/share/classes/java/lang/String.java:2506]: While not directly observable (all tests pass) then sadly the intrinsic implementation emits a string range check that triggers a decompilation and subsequent compilation without the intrinsic if you'd call with a `fromIndex` with values greater than `length()` (equal to should be fine). I'll try out some options here.

**Comment by cl4es** [comment]
Updated to clamp `fromIndex` to `value.length >> coder` since the intrinsic might silently deoptimize if passed a `fromIndex` value that's out of bounds.

This correctness fix dials back the speed-up on some of the `indexOf(c, offset)` micros, but not by much:
```
Name                                               Cnt   Base   Error    Test   Error  Unit  Change
StringIndexOf.searchChar16LongSuccess             5/25 14,909 ± 3,284  14,351 ± 0,105 ns/op   1,04x (p = 0,218 )
StringIndexOf.searchChar16LongWithOffsetSuccess   5/25 14,715 ± 0,870  14,098 ± 0,015 ns/op   1,04x (p = 0,004*)
StringIndexOf.searchChar16MediumSuccess           5/25  6,611 ± 0,016   6,488 ± 0,025 ns/op   1,02x (p = 0,000*)
StringIndexOf.searchChar16MediumWithOffsetSuccess 5/25  7,151 ± 0,026   7,064 ± 0,021 ns/op   1,01x (p = 0,000*)
StringIndexOf.searchChar16ShortSuccess            5/25  2,667 ± 0,007   2,362 ± 0,002 ns/op   1,13x (p = 0,000*)
StringIndexOf.searchChar16ShortWithOffsetSuccess  5/25  2,494 ± 0,112   2,543 ± 0,008 ns/op   0,98x (p = 0,018 )
StringIndexOf.searchCharLongSuccess               5/25  5,995 ± 0,017   5,270 ± 0,005 ns/op   1,14x (p = 0,000*)
StringIndexOf.searchCharLongWithOffsetSuccess     5/25  6,377 ± 0,118   6,212 ± 0,003 ns/op   1,03x (p = 0,000*)
StringIndexOf.searchCharMediumSuccess             5/25  2,350 ± 0,013   1,924 ± 0,002 ns/op   1,22x (p = 0,000*)
StringIndexOf.searchCharMediumWithOffsetSuccess   5/25  2,675 ± 0,002   2,581 ± 0,005 ns/op   1,04x (p = 0,000*)
StringIndexOf.searchCharShortSuccess              5/25  1,719 ± 0,002   1,243 ± 0,002 ns/op   1,38x (p = 0,000*)
StringIndexOf.searchCharShortWithOffsetSuccess    5/25  1,635 ± 0,006   1,311 ± 0,001 ns/op   1,25x (p = 0,000*)
  * = significant
```

**Review by Hamlin-Li — APPROVED** [review]
Looks good!

**Review by RogerRiggs — APPROVED** [review]

**Comment by cl4es** [comment]
Thanks for the reviews!

/integrate

**Pull request closed without merging**